### ISSUE TITLE
error: Problem running ghostscript gs -q -g562x56 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps. Check your installation!

### ISSUE THREAD

**Issue by malaterre** [doxygen/doxygen#7290]
Hi

I am trying to build the GDCM documentation using doxygen from Debian package. It currently report an error (somewhere in the header file).

Error is:

```
Last OS error: Permission denied
GPL Ghostscript RELEASE CANDIDATE 3 9.28: Unrecoverable error, exit code 1
error: Problem running ghostscript gs -q -g562x56 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps. Check your installation!
```

Executing locally I get:

```
$ gs -q -g562x56 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps
Error: /undefinedfilename in (_form0.ps)
Operand stack:

Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push
Dictionary stack:
   --dict:729/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--
Current allocation mode is local
Last OS error: Permission denied
GPL Ghostscript RELEASE CANDIDATE 3 9.28: Unrecoverable error, exit code 1
```

**Comment by albert-github** [comment]
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.
- Please also specify the doxygen version used.

**Comment by malaterre** [comment]
Where does this `_form0.ps` comes from ? Using a somewhat verbose doxygen I see:

```
Generating docs for compound gdcm::network::ServiceClassApplicationInformation...
GenerThis is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/Debian) (preloaded format=latex)
 restricted \write18 enabled.
entering extended mode
(./_formulas.tex
LaTeX2e <2018-12-01>

Error: /undefinedfilename in (_form0.ps)
Operand stack:

Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   
false   1   %stopped_push
Dictionary stack:
   --dict:729/1123(ro)(G)--   --dict:0/20(G)--   --dict:76/200(L)--
Current allocation mode is local
Last OS error: Permission denied
GPL Ghostscript RELEASE CANDIDATE 3 9.28: Unrecoverable error, exit code 1
error: Problem running ghostscript gs -q -g562x56 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps. Check your installation!
ating docs for compound gdcm::network::SOPClassExtendedNegociationSub...

```

**Comment by malaterre** [comment]
```
$ doxygen --version
1.8.13
```

**Comment by malaterre** [comment]
@gerddie

**Comment by malaterre** [comment]
@albert-github How can I extract a small example ? This is failing right in the middle of the generating of documentation for GDCM:

* https://github.com/malaterre/GDCM

**Comment by malaterre** [comment]
@albert-github If that help:

```
$ gs --help | grep ppm 
   png48 pngalpha pnggray pngmono pngmonod pnm pnmraw ppm ppmraw pr1000

```

**Comment by albert-github** [comment]
The current doxygen version is 1.8.16 can you have a try with this version?
There are some small improvements in the master version regarding formulas as well so maybe it is even better to try the master version.

It looks like the problem is in the the generation of the images of the formulas for the HTML output (nicer examples might be to use MathJax but this is not the question here).
Generating a small example is, probably, reasonable easy, on the HTML directory there is the file: _formulas.tex in this are the formulas to be "translated" and in this it is the first formula (normally the file is removed but it is probably present as the conversion process did not succeed). with the text of this formula you should be able to find the source file giving the problem and create a small example from there.

**Comment by malaterre** [comment]
[doxygen.zip](https://github.com/doxygen/doxygen/files/3681413/doxygen.zip)

**Comment by malaterre** [comment]
@albert-github Thanks I see it now:


```
$ cat _formulas.tex
\batchmode
\documentclass{article}
\usepackage{epsfig}
\pagestyle{empty}
\begin{document}
\[ RWV = 1. * SV - 1024 \]
\pagebreak

\end{document}

```

So the header problem is here:

* https://github.com/malaterre/GDCM/blob/master/Source/MediaStorageAndFileFormat/gdcmRescaler.h#L35

**Comment by albert-github** [comment]
This looks at first glance to be the file that is provoking the error, but I don't think you included the mentioned file in you example. 

I tried, under Cygwin, with a 1.8.13 executable to reproduce the problem but I was not able to, the output nicely shows the image of the formula.
- I took the file gdcmRescaler.h  
- created a default Doxyfile with version 1.8.13
- ran doxyen.

Maybe you can create a new minimal example wit just the mentioned files as well as the html directory, so we are 100% sure that we both use the same example..

I see that you ran with TeX Live 2019/Debian, I'm using a 2018 version might be a small difference (but I don't expect it in this case).
- what does `gs --version` give
- what does `which gs` give
- when running doxygen, please run with `doxygen -d extcmd` and report the lines starting with "Executing external command" or better include the entire terminal output into the new minimal example.

**Comment by albert-github** [comment]
On my default system I use gs 9.23 on another system I use gs 9.27, according to the information I saw you are using GPL Ghostscript RELEASE CANDIDATE 3 9.28. Maybe you can check with an older version (e.g. 9.27) from https://www.ghostscript.com/download/gsdnld.html or try the newest release candidate (From the information  at http://git.ghostscript.com/?p=ghostpdl.git;a=summary it looks to me that that RC4 is available).

**Comment by vmatare** [comment]
Any idea what's triggering this or what the actual problem is? Because I'm now also running into it with ghostscript-9.28_rc4. Downgrading to 9.26 is a workaround. Maybe this needs to be reported upstream?

**Comment by albert-github** [comment]
I have no idea what triggered this problem, but as the older versions do work and the ghostscript-9.28_rc4 doesn't I think it would be good to report the problem upstream and in the mean time downgrade.

**Comment by marehr** [comment]
This also broke for me.

**Comment by albert-github** [comment]
@marehr Which OS?

Can you please attach a, small, self contained example (source+configuration file+output in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.

**Comment by marehr** [comment]
arch linux with most recent updates which include gs-9.50 and doxygen-1.8.16.

I don't know if it is a regression in ghostscript or if your output file is wrong.

[test.zip](https://github.com/doxygen/doxygen/files/3831708/test.zip) is a really small example with a single header file that includes formula + all the generated output is included as well

**Comment by albert-github** [comment]
I did some tests with the 9.50 version (under Windows) and I have the same problems, with the older versions (that I have 9.05 and 9.23) there are no problems.

**Comment by marehr** [comment]
@albert-github Do you know where or how the `_form0.ps` file is generated? I wanted to file a regression report at https://bugs.ghostscript.com/, but it might be that this generated ps file was always ill-formed.

**Comment by albert-github** [comment]
It might be that the file is ill-formed in that case it is interesting to know what the problem is, it is more likely that `_form0.eps` is ill_formed, but that file is a generated file (but still useful to know what is wrong).
The file `_form0.ps` is generated in the HTML directory along with the file `_form0.eps`. `_form0.eps` is run from `_form0.ps`.

**Comment by marehr** [comment]
> `_form0.eps` is run from `_form0.ps`.

So `_form0.ps` is the source for all of it? I see an eps file in the directory and it can be opened and it seems to be valid.

So the transformation is:

* generate _form0.eps: `some_tool _form0.ps > _form0.eps` (do you know the tool? I tried [ps2eps](http://manpages.ubuntu.com/manpages/trusty/man1/ps2eps.1.html) and it didn't work)
* generate _form0.pnm: `gs -q -g178x66 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps`

But, how is _form0.ps created?

**Comment by albert-github** [comment]
The basic information is in the log file when running doxygen with the `-d extcmd`:
- Executing external command `latex _formulas.tex`
   Generating a `_formulas.dvi` file based on the formulas in the code and written by doxygen to `_formulas.tex`
- Executing external command `dvips -q -D 600 -E -n 1 -p 1 -o _form0.eps _formulas.dvi`
  converting the `dvi` file to `eps` by means of `dvips` (here the first page / formula)
- Not shown! as external command as the `_form0.ps` is written by doxygen based on the file name and the BoundingBox in the `_form0.eps`
- Executing external command `gswin32c.exe -q -g178x66 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps`
  generating from the `_form0.ps` and the `_form0.eps` the file `_form0.pnm` that is later on read by doxygen to translate into a png file.

**Comment by marehr** [comment]
Okay thank you for outline, this seems to be the code:

https://github.com/doxygen/doxygen/blob/4fbb61f020a607a96152ba5c00668b9840ea812c/src/formula.cpp#L164-L184

**Comment by albert-github** [comment]
I did some more experiments by running so9me versions:
- version 9.27: works
- version 9.28 rc1: doesn't work, has the problems we see

9.28 is the working version based on which 9.50 is released.

**Comment by albert-github** [comment]
Indeed the mentioned section is the part that generates the `ps` file, the line where it goes wrong in the execution looks like the line that is written by ` t << "(" << formBase << ".eps) run" << endl; `

I think the people from ghostscript should be able to give an answer based on the `ps` and `eps` file as well as the used command: `gswin32c.exe -q -g178x66 -r384x384x -sDEVICE=ppmraw -sOutputFile=_form0.pnm -dNOPAUSE -dBATCH -- _form0.ps`

Please post the link when you created a bug report.

**Comment by marehr** [comment]
Yeah, opened a regression question at https://bugs.ghostscript.com/show_bug.cgi?id=701894

**Comment by albert-github** [comment]
Best is also to post the `eps` file as this is used / referenced.

**Comment by marehr** [comment]
I had a talk with Robin Watts and Ken Sharp at IRC and there seem to be basically 3 different problems: 
* `-r%dx%d` (the dimension for `r` shouldn't be `-r384x384x`, but `-r384x384`),
* misuse / unnecessary use of `--` and
* since 9.50 the command needs more control access (that might be worked around by either whitelisting the file via `--permit-file-read=_form0.eps` (only works from 9.50 and upwards) or generally accepting any file with `-dNOSAFER` (works since quite some time). The second option is considered to be unsafe if we would process any file, but in this case we process self produced / controlled files. I don't know if doxygen has any threat model that it assumes. ). 

> Ken Sharp: Yeah the %dx is wrong, as Robin says its sheer luck that works
the -- isn't needed and is what's causing the first problem
and file control is the new bugbear

**Comment by albert-github** [comment]
@marehr, Robin Watts and Ken Sharp Thanks for discussing this issue and finding a solution.

I've just pushed a proposed patch, pull request #7392

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
